cmd/regres/main.go

```diff
@@ -0,0 +1,239 @@
+	app.Run(run)
+}
+
+type stats struct {
+	SHA                  string
+	BuildTime            float64  // in seconds
+	IncrementalBuildTime float64  // in seconds
+	FileSizes            struct { // in bytes
+		LibGAPII                   int64
+		LibVkLayerVirtualSwapchain int64
+		GAPIDAarch64APK            int64
+		GAPIDArmeabi64APK          int64
+		GAPIDX86APK                int64
+		GAPID                      int64
+		GAPIR                      int64
+		GAPIS                      int64
+		GAPIT                      int64
+	}
+}
+
+func run(ctx context.Context) error {
+	if *root == "" {
+		wd, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		*root = wd
+	}
+
+	g, err := git.New(*root)
+	if err != nil {
+		return err
+	}
+	s, err := g.Status(ctx)
+	if err != nil {
+		return err
+	}
+	if !s.Clean() {
+		return fmt.Errorf("Local changes found. Please submit any changes and run again")
+	}
+
+	branch, err := g.CurrentBranch(ctx)
+	if err != nil {
+		return err
+	}
+
+	defer g.CheckoutBranch(ctx, branch)
+
+	cls, err := g.Log(ctx, *count)
+	if err != nil {
+		return err
+	}
+
+	rnd := rand.New(rand.NewSource(time.Now().Unix()))
+
+	res := []stats{}
+	for i := range cls {
+		i := len(cls) - 1 - i
+		cl := cls[i]
+		sha := cl.SHA.String()[:6]
+
+		r := stats{SHA: sha}
+
+		log.I(ctx, "HEAD~%.2d: Building at %v: %v", i, sha, cl.Subject)
+		if err := g.Checkout(ctx, cl.SHA); err != nil {
+			return err
+		}
+
+		duration, err := build(ctx)
+		if err != nil {
+			continue
+		}
+		r.BuildTime = duration.Seconds()
+
+		if err := withTouchedGLES(ctx, rnd, func() error {
+			log.I(ctx, "HEAD~%.2d: Building incremental change at %v: %v", i, sha, cl.Subject)
+			if duration, err := build(ctx); err == nil {
+				r.IncrementalBuildTime = duration.Seconds()
+			}
+			return nil
+		}); err != nil {
+			continue
+		}
+
+		pkg := filepath.Join(*root, "bazel-bin", "pkg")
+		for _, f := range []struct {
+			path string
+			size *int64
+		}{
+			{filepath.Join(pkg, "lib", dllExt("libgapii")), &r.FileSizes.LibGAPII},
+			{filepath.Join(pkg, "lib", dllExt("libVkLayer_VirtualSwapchain")), &r.FileSizes.LibVkLayerVirtualSwapchain},
+			{filepath.Join(pkg, "gapid-aarch64.apk"), &r.FileSizes.GAPIDAarch64APK},
+			{filepath.Join(pkg, "gapid-armeabi.apk"), &r.FileSizes.GAPIDArmeabi64APK},
+			{filepath.Join(pkg, "gapid-x86.apk"), &r.FileSizes.GAPIDX86APK},
+			{filepath.Join(pkg, exeExt("gapid")), &r.FileSizes.GAPID},
+			{filepath.Join(pkg, exeExt("gapir")), &r.FileSizes.GAPIR},
+			{filepath.Join(pkg, exeExt("gapis")), &r.FileSizes.GAPIS},
+			{filepath.Join(pkg, exeExt("gapit")), &r.FileSizes.GAPIT},
+		} {
+			fi, err := os.Stat(f.path)
+			if err != nil {
+				continue
+			}
+			*f.size = fi.Size()
+		}
+		res = append(res, r)
+	}
+
+	fmt.Printf("-----------------------\n")
+
+	w := tabwriter.NewWriter(os.Stdout, 1, 4, 0, ' ', 0)
+	defer w.Flush()
+
+	fmt.Fprint(w, "sha"+
+		"\t | build_time"+
+		"\t | incremental_build_time"+
```

Comment:
nit: `build_time` should be called `incremental_build_time`, since that's what it really is, while `incremental_build_time` should be called something like `gles_api_build_time`.

Reply:
Well, yes they're both incremental build times. I'm not sure that there's too much value in `build_time` so I might just remove it. Going to be adding more stuff to this tool, so I'd like to follow up with more CLs to fix this.